[WTF] HashTable::rehash should not perform key comparison

Constellation opened the pull request:
#### 4e0aad18e7ef1a98520e937dab3e5fb62131564e
<pre>
[WTF] HashTable::rehash should not perform key comparison
<a href="https://bugs.webkit.org/show_bug.cgi?id=242628">https://bugs.webkit.org/show_bug.cgi?id=242628</a>

Reviewed by Darin Adler.

When rehashing a table from old one to new one, there are no duplicate keys since it is HashTable.
Also, there is no deleted keys in the new HashTable. We changed lookupForWriting to lookupForReinsert
so that this function looks for an insertion point by only checking whether a hashtable entry is empty.
We also use it for addUniqueForInitialization.

* Source/WTF/wtf/HashTable.h:
(WTF::HashTable::lookupForReinsert):
(WTF::KeyTraits&gt;::lookupForReinsert):
(WTF::KeyTraits&gt;::addUniqueForInitialization):
(WTF::KeyTraits&gt;::reinsert):
(WTF::HashTable::lookupForWriting): Deleted.
(WTF::KeyTraits&gt;::lookupForWriting): Deleted.

Canonical link: <a href="https://commits.webkit.org/252386@main">https://commits.webkit.org/252386@main</a>
</pre>

darinadler commented on Source/WTF/wtf/HashTable.h:
```diff
@@ -1051,16 +999,16 @@ DECLARE_ALLOCATOR_WITH_HEAP_IDENTIFIER(HashTable);
-        ASSERT(!lookupForWriting(Extractor::extract(entry)).second);
-        ASSERT(!isDeletedBucket(*(lookupForWriting(Extractor::extract(entry)).first)));
+        ASSERT(!isDeletedBucket(*(lookupForReinsert(Extractor::extract(entry)))));
+        ASSERT(isEmptyBucket(*(lookupForReinsert(Extractor::extract(entry)))));
```
I would have omitted the parentheses after the `*` here. I think they actually make the code slightly harder to read.

Constellation replied:
Fixed.

darinadler commented on Source/WTF/wtf/HashTable.h:
```diff
@@ -542,9 +542,9 @@ DECLARE_ALLOCATOR_WITH_HEAP_IDENTIFIER(HashTable);
         typedef std::pair<LookupType, unsigned> FullLookupType;
 
-        LookupType lookupForWriting(const Key& key) { return lookupForWriting<IdentityTranslatorType>(key); };
+        ValueType* lookupForReinsert(const Key& key) { return lookupForReinsert<IdentityTranslatorType>(key); };
```
Change Log comment says "only checks whether it is an empty." That’s a sentence fragment, missing a word.

Constellation replied:
Changed it to `looks for an insertion point by only checking whether a hashtable entry is empty`.

Constellation committed:
[WTF] HashTable::rehash should not perform key comparison
https://bugs.webkit.org/show_bug.cgi?id=242628

Reviewed by Darin Adler.

When rehashing a table from old one to new one, there are no duplicate keys since it is HashTable.
Also, there is no deleted keys in the new HashTable. We changed lookupForWriting to lookupForReinsert
so that this function looks for an insertion point by only checking whether a hashtable entry is empty.
We also use it for addUniqueForInitialization.

* Source/WTF/wtf/HashTable.h:
(WTF::HashTable::lookupForReinsert):
(WTF::KeyTraits>::lookupForReinsert):
(WTF::KeyTraits>::addUniqueForInitialization):
(WTF::KeyTraits>::reinsert):
(WTF::HashTable::lookupForWriting): Deleted.
(WTF::KeyTraits>::lookupForWriting): Deleted.

Canonical link: https://commits.webkit.org/252386@main
